pipelinerl/async_llm.py

```diff
@@ -1,16 +1,55 @@
+import base64
+import io
 import logging
+
 import aiohttp
+import numpy as np
+from PIL import Image
+from tapeagents.core import LLMCall, LLMOutput, Prompt, TokenLogprob
+from tapeagents.llms.trainable import TrainableLLM
 
 from pipelinerl.finetune.data import MASKED_TOKEN_ID
+from pipelinerl.rollouts import TrainingText
+from pipelinerl.processor_factory import processor_factory
 
-from tapeagents.core import LLMCall, LLMOutput, Prompt, TokenLogprob, TrainingText
-from tapeagents.llms.trainable import TrainableLLM
+logger = logging.getLogger(__name__)
 
 
-logger = logging.getLogger(__name__)
+def extract_images_from_messages(messages: list[dict]) -> list[Image.Image]:
+    """Extract PIL Images from multimodal messages."""
+
+    images = []
+    for message in messages:
+        if isinstance(message.get("content"), list):
+            for content_item in message["content"]:
+                if content_item is None:
+                    continue
+                if content_item.get("type") == "image" and "image" in content_item:
+                    images.append(content_item["image"])
+                elif (
+                    content_item.get("type") == "image_url"
+                    and "image_url" in content_item
+                ):
+                    # Handle base64 format
+                    url = content_item["image_url"]["url"]
+                    if url.startswith("data:image;base64,"):
+                        try:
+                            base64_data = url.split("data:image;base64,")[1]
+                            image_data = base64.b64decode(base64_data)
+                            image = Image.open(io.BytesIO(image_data))
+                            images.append(image)
+                        except Exception as e:
+                            raise e
+
+    if not images:
+        raise ValueError("No images found in messages")
 
+    return images
 
-async def llm_async_generate(llm: TrainableLLM, prompt: Prompt, session: aiohttp.ClientSession) -> LLMCall:
+
+async def llm_async_generate(
```

Comment:
Can we add a deprecation comment here? I think most people should use `TrainableLLM.agenerate`  from TapeAgents directly.